src/org/sosy_lab/java_smt/example/ListAllSolversWithFeatures.java

```diff
@@ -0,0 +1,311 @@
+    listAllSolversWithFeatures.printGridToStdOut(infoArray);
+  }
+
+  public ListAllSolversWithFeatures() {}
+
+  /**
+   * Prints the infoArray to StdOut in a nice table using the formating method (through printf()).
+   * Note that this conforms to the row/columns specified in getInformationForAllSolvers().
+   */
+  private void printGridToStdOut(List<List<String>> infoArray) {
+    if (infoArray == null || infoArray.get(0).size() == 0) {
+      System.out.println("Could not find any installed SMT-Solvers.");
+      return;
+    }
+
+    int maxSolverLength = lineLength(infoArray, 0);
+    int maxVersionLength = lineLength(infoArray, 1);
+    int maxTheoriesLength = lineLength(infoArray, 2);
+    int maxFeaturesLength = lineLength(infoArray, 3);
+    String infoColumn =
+        "| %-"
+            + maxSolverLength
+            + "s | %-"
+            + maxVersionLength
+            + "s | %-"
+            + maxTheoriesLength
+            + "s | %-"
+            + maxFeaturesLength
+            + "s |%n";
+    String seperatorColumn =
+        "+-"
+            + "-".repeat(maxSolverLength)
+            + "-+-"
+            + "-".repeat(maxVersionLength)
+            + "-+-"
+            + "-".repeat(maxTheoriesLength)
+            + "-+-"
+            + "-".repeat(maxFeaturesLength)
+            + "-+%n";
+
+    System.out.printf(seperatorColumn);
+    System.out.printf(infoColumn, "Solver", "Version", "Theories", "Features");
+    System.out.printf(seperatorColumn);
+
+    for (List<String> s : infoArray) {
+      System.out.printf(infoColumn, s.get(0), s.get(1), s.get(2), s.get(3));
+    }
+
+    System.out.printf(seperatorColumn);
+  }
+
+  /**
+   * Gathers information about every available solver into a 2-Dimensional array representing rows
+   * and columns of the later table. Columns represent (in order): Solver, Version, Theories,
+   * Features. Rows represent the solvers available to the installed JavaSMT.
+   */
+  private List<List<String>> getInformationForAllSolvers()
+      throws SolverException, InterruptedException {
+    List<List<String>> info = new ArrayList<>();
+    for (Solvers s : Solvers.values()) {
+      List<String> solverInfo = getSolverInformation(s);
+      if (solverInfo != null && solverInfo.size() == 4) {
+        info.add(solverInfo);
+      }
+    }
+    return info;
+  }
+
+  /**
+   * Checks for solver-name, version, theories and features and packs them into a String[] of length
+   * 4.
+   *
+   * @param solver to check for information. Taken from Solvers enum only.
+   * @return String[] of length 4 of the solver you entered. Array content as String in the
+   *     following order: SolverName, SolverVersion, SolverTheories, SolverFeatures. String[] length
+   *     0 if invalid solver.
+   */
+  private List<String> getSolverInformation(Solvers solver)
+      throws SolverException, InterruptedException {
+    List<String> info = new ArrayList<>();
```

Comment:
An `info` object with named members would be nicer .... perhaps.

Reply:
`info` object in a Class?
Could you explain further please?

Reply:
> The table is quite nice and clearly written, I like it. It is just very broad.
> A possible solution would be to use multiple lines per cell, but that might be quite tricky to implement in a simple and understandable way.
Yes i know. But i know that its possible to "unlock" the terminal/shell so that you can scroll and that there are no forced line breaks, i just dont know how yet.

> And perhaps, we should aim for a more object-oriented approach instead of using lists of lists of Strings.
Done

Reply:
The SolverInfo class would allow to split data retrieval and data dumping. Just an idea.
The following draft might help:
```
/** data class containing the info for the solver and can print it */
class SolverInfo {
  - Solvers solver;
  - String version;
  - List<String> supportedTheories;
  - List<String> features;
}
/** helper class for wrapping the contents of a row in the table  */
class Row {
  - List<String> lines;
  - int minNumOfLines;
  - List<Integer> columnWidths; // needed for producing a new empty line with seperators "|   |   |   |"
  + appendCell(List<String> lines); // append cell-lines to lines. Add new line, if number of cell lines is too large.   
  + toString(); // dump nice row, include terminal line as last line: "|---|---|---|"
}
+ List<String> cellToRow(String cellContent, int exactLengthOfCell); // returns a list of lines, formatted to match exact length or less space, does no contain seperator "|"
```